### PR TITLE
Bug fixes in #88

### PR DESCRIPTION
**Pull request by tgsong**
*Description of changes:*

- Fix include of stdbool.h
- decodeFileBlock should accept const network buffer
- decodeFileBlock_mqtt should not reuse network buffer
- remove unused methods
- signature buffer should not be dynamically allocated
- only update http url when it's not NULL
- decoded data payload should be freed after processDataBlock

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: source/ota.c [hunk 1]
```diff
@@ -1262,45 +1155,34 @@ static DocParseErr_t decodeAndStoreKey( char * pValueInJson,
...
+                    "base64Decode returned error: "
+                    "error=%d",
+                    base64Status ) );
+        err = DocParseErrBase64Decode;
     }
     else
     {
-        /* We failed to allocate needed memory. Everything will be freed below upon failure. */
-        err = DocParseErrOutOfMemory;
+        ( *pSig256 )->size = ( uint16_t ) actualLen;
```
Comment (yanjos-dev) [inline source/ota.c:1182]: Is this case not considered an error? Also, if actualLen is always supposed to be zero here, then we may just want to assign this to zero directly. This way it's more explicit that this is the behavior during an error.

Comment (tgsong) [reply source/ota.c:1182]: What do you mean? This is the success path right? `base64Status == Base64Success` in this else branch.

File: source/ota.c [hunk 2]
```diff
@@ -1789,15 +1671,20 @@ static OtaJobParseErr_t verifyActiveJobStatus( OtaFileContext_t * pFileContext,
...
+                {
+                    /* The buffer is allocated by us, free first then update. */
+                    otaAgent.pOtaInterface->os.mem.free( otaAgent.fileContext.pUpdateUrlPath );
+                    otaAgent.fileContext.pUpdateUrlPath = pFileContext->pUpdateUrlPath;
+                    pFileContext->pUpdateUrlPath = NULL;
+                }
+                else
+                {
+                    /* The buffer is provided by user, directly copy the new url to it. */
+                    memcpy( otaAgent.fileContext.pUpdateUrlPath, pFileContext->pUpdateUrlPath, otaAgent.fileContext.updateUrlMaxSize );
```
Comment (yanjos-dev) [inline source/ota.c:1686]: Somewhat unrelated to this PR, so it could be considered in a separate PR, but shouldn't we be checking that the amount of data we are trying to copy is less than the max amount we can store? I see that it won't be possible to write outside the bounds of the array, but we may not be copying the complete URL.

**Review by pvyawaha — APPROVED** [review]

File: source/ota.c [hunk 3]
```diff
@@ -1262,45 +1155,34 @@ static DocParseErr_t decodeAndStoreKey( char * pValueInJson,
...
 {
     DocParseErr_t err = DocParseErrNone;
+    size_t actualLen = 0;
+    Base64Status_t base64Status = 0;
+    Sig256_t ** pSig256 = pParamAdd;
 
-    /* Allocate space for and decode the base64 signature. */
-    void * pSignature = otaAgent.pOtaInterface->os.mem.malloc( sizeof( Sig256_t ) );
+    /* pSig256 should point to pSignature in OtaFileContext_t, which is statically allocated. */
+    assert( *pSig256 != NULL );
```
Comment (yanjos-dev) [inline source/ota.c:1163]: Is there an upper bound for the size of a signature that we can receive in a job document? I didn't think that there was.

File: source/ota.c (hunk 2 above)
Comment (tgsong) [reply source/ota.c:1686]: destination is `otaAgent.fileContext.pUpdateUrlPath`, and size we're copying is `otaAgent.fileContext.updateUrlMaxSize`, which is the buffer size of that destination pointer. `pFileContext->pUpdateUrlPath` I believe is pointing to network buffer.

File: source/ota.c (hunk 3 above)
Comment (tgsong) [reply source/ota.c:1163]: See `kOTA_MaxSignatureSize`, it's part of `Sig256_t` struct.

File: source/ota.c (hunk 1 above)
Comment (yanjos-dev) [reply source/ota.c:1182]: You're right, I misread the condition
